include/picongpu/fields/currentDeposition/Esirkepov/EsirkepovNative.hpp

```diff
@@ -34,182 +33,195 @@ namespace picongpu
-                                           velocity.y() * deltaTime / cellSize.y(),
-                                           velocity.z() * deltaTime / cellSize.z());
-        const PosType oldPos = pos - deltaPos;
-        Line<float3_X> line(oldPos, pos);
-        auto cursorJ = dataBoxJ.toCursor();
-
-        /**
-         * \brief the following three calls separate the 3D current deposition
-         * into three independent 1D calls, each for one direction and current component.
-         * Therefore the coordinate system has to be rotated so that the z-direction
-         * is always specific.
+        using ParticleAssign = typename T_ParticleShape::ChargeAssignment;
+        static constexpr int supp = ParticleAssign::support;
+
+        static constexpr int currentLowerMargin = supp / 2 + 1;
+        static constexpr int currentUpperMargin = (supp + 1) / 2 + 1;
+        typedef pmacc::math::CT::Int<currentLowerMargin, currentLowerMargin, currentLowerMargin> LowerMargin;
+        typedef pmacc::math::CT::Int<currentUpperMargin, currentUpperMargin, currentUpperMargin> UpperMargin;
+
+        PMACC_CASSERT_MSG(
+            __EsirkepovNative_supercell_or_number_of_guard_supercells_is_too_small_for_stencil,
+            pmacc::math::CT::min<
+                typename pmacc::math::CT::mul<
+                    SuperCellSize,
+                    GuardSize
+                >::type
+            >::type::value >= currentLowerMargin &&
+            pmacc::math::CT::min<
+                typename pmacc::math::CT::mul<
+                    SuperCellSize,
+                    GuardSize
+                >::type
+            >::type::value >= currentUpperMargin
+        );
+
+        /* iterate over all grid points */
+        static constexpr int begin = -currentLowerMargin;
+        static constexpr int end = currentUpperMargin + 1;
+
+        float_X charge;
+
+        /* At the moment Esirkepov only supports Yee cells where W is defined at origin (0,0,0)
+         *
+         * \todo: please fix me that we can use CenteredCell
          */
+        template<
+            typename DataBoxJ,
+            typename PosType,
+            typename VelType,
+            typename ChargeType,
+            typename T_Acc
+        >
+        DINLINE void operator()(
+            T_Acc const & acc,
+            DataBoxJ dataBoxJ,
+            const PosType pos,
+            const VelType velocity,
+            const ChargeType charge, const float_X deltaTime
+        )
+        {
+            this->charge = charge;
+            const float3_X deltaPos = float3_X(velocity.x() * deltaTime / cellSize.x(),
+                                               velocity.y() * deltaTime / cellSize.y(),
+                                               velocity.z() * deltaTime / cellSize.z());
+            const PosType oldPos = pos - deltaPos;
+            Line<float3_X> line(oldPos, pos);
```

Comment:
Uh, I think that can be `const`:
```suggestion
            const Line<float3_X> line(oldPos, pos);
```